### PR TITLE
datetime: add TZ support

### LINKED ISSUE #163: Support TZ in datetime
- Support datetime #118
- Issue in Tarantool - "datetime: Olson/timezones database support"  https://github.com/tarantool/tarantool/issues/6751

### PR DESCRIPTION
**Pull request by oleg-jukovec**
The patch adds timezone support [1] for datetime. For the purpose of compatibility we got constants for timezones from Tarantool [2]. The patch adds a script to generate the data according to the instructions [3].

1. https://github.com/tarantool/tarantool/wiki/Datetime-Internals#timezone-support
2. https://github.com/tarantool/tarantool/blob/9ee45289e01232b8df1413efea11db170ae3b3b4/src/lib/tzcode/timezones.h
3. https://github.com/tarantool/tarantool/blob/9ee45289e01232b8df1413efea11db170ae3b3b4/src/lib/tzcode/gen-zone-abbrevs.pl#L35-L39

I didn't forget about (remove if it is not applicable):

- [x] Tests (see [documentation](https://pkg.go.dev/testing) for a testing package)
- [x] Changelog (see [documentation](https://keepachangelog.com/en/1.0.0/) for changelog format)
- [x] Documentation (see [documentation](https://go.dev/blog/godoc) for documentation style guide)

Related issues:

Closes #163

### CODE REVIEW

File: datetime/datetime_test.go [hunk 1]
```diff
@@ -565,6 +698,31 @@ func TestMPDecode(t *testing.T) {
...
+	}
+	if err.Error() != "Invalid data length: got 1, wanted 8 or 16" {
+		t.Fatalf("Unexpected error: %s", err.Error())
+	}
+}
+
+func TestUnmarshalMsgpackInvalidZone(t *testing.T) {
+	var v Datetime
+
+	buf, _ := hex.DecodeString("b016b9430000000000000000b400ee01")
```
Comment (DifferentialOrange) [inline datetime/datetime_test.go:716]: I think it's worth do add a comment about what's been encoded here.

File: datetime/example_test.go [hunk 2]
```diff
@@ -79,3 +79,24 @@ func Example() {
...
+// timezone in datetime.
+func ExampleNewDatetime_noTimezone() {
+	var datetime = "2013-10-28T17:51:56.000000009Z"
+	tm, err := time.Parse(time.RFC3339, datetime)
+	if err != nil {
+		fmt.Printf("error in time.Parse() is %v", err)
+		return
+	}
+
+	tm = tm.In(NoTimezone)
```
Comment (DifferentialOrange) [inline datetime/example_test.go:93]: So now I couldn't create a datetime without specifying timezone: I should either set a valid timezone or set NoTimezone timezone?

File: datetime/export_test.go [hunk 3]
```diff
@@ -0,0 +1,5 @@
+package datetime
```
Comment (DifferentialOrange) [inline datetime/export_test.go:1]: Shouldn't it be some `datetime_test` or it won't work?

**Review by DifferentialOrange — APPROVED** [review]
Seems awesome.

I have left a couple of questions

File: datetime/datetime_test.go (hunk 1 above)
Comment (oleg-jukovec) [reply datetime/datetime_test.go:723]: Thank you, added:
```patch
@@ -713,6 +716,10 @@ func TestUnmarshalMsgpackInvalidLength(t *testing.T) {
 func TestUnmarshalMsgpackInvalidZone(t *testing.T) {
        var v Datetime
 
+       // The original value from datetimeSample array:
+       // {time.RFC3339 + " MST",
+       //  "2006-01-02T15:04:00+03:00 MSK",
+       //  "d804b016b9430000000000000000b400ee00"}
```

File: datetime/datetime_test.go [hunk 4]
```diff
@@ -73,6 +74,125 @@ func assertDatetimeIsEqual(t *testing.T, tuples []interface{}, tm time.Time) {
...
+func TestInvalidTimezone(t *testing.T) {
+	invalidLoc := time.FixedZone("AnyInvalid", 0)
+	tm, err := time.Parse(time.RFC3339, "2010-08-12T11:39:14Z")
+	if err != nil {
+		t.Fatalf("Time parse failed: %s", err)
+	}
+	tm = tm.In(invalidLoc)
+	dt, err := NewDatetime(tm)
+	if err == nil {
+		t.Fatalf("Unexpected success: %v", dt)
```
[the remark this answers is not in the document]
Comment (oleg-jukovec) [reply datetime/datetime_test.go:102]: It's a great idea. Updated:
```patch
--- a/datetime/datetime_test.go
+++ b/datetime/datetime_test.go
@@ -101,6 +101,9 @@ func TestInvalidTimezone(t *testing.T) {
        if err == nil {
                t.Fatalf("Unexpected success: %v", dt)
        }
+       if err.Error() != "Unknown timezone AnyInvalid with offset 0" {
+               t.Fatalf("Unexpected error: %s", err.Error())
+       }
 }
 
 func TestInvalidOffset(t *testing.T) {
 ```

File: datetime/datetime.go [hunk 5]
```diff
@@ -105,8 +128,12 @@ func (dtime *Datetime) MarshalMsgpack() ([]byte, error) {
 	var dt datetime
 	dt.seconds = tm.Unix()
 	dt.nsec = int32(tm.Nanosecond())
-	dt.tzIndex = 0  // It is not implemented, see gh-163.
-	dt.tzOffset = 0 // It is not implemented, see gh-163.
+
+	zone, offset := tm.Zone()
+	if zone != noTimezoneZone {
+		dt.tzIndex = int16(timezoneToIndex[zone])
```
[the remark this answers is not in the document]
Comment (oleg-jukovec) [reply datetime/datetime.go:136]: The valid ways to create Datetime is NewDatetime() or UnmatshalMsgpack() calls. We already check the values there.

I've added a comment to the code:
```patch
--- a/datetime/datetime.go
+++ b/datetime/datetime.go
@@ -131,6 +131,8 @@ func (dtime *Datetime) MarshalMsgpack() ([]byte, error) {
 
        zone, offset := tm.Zone()
        if zone != noTimezoneZone {
+               // The zone value already checked in NewDatetime() or
+               // UnmarshalMsgpack() calls.
                dt.tzIndex = int16(timezoneToIndex[zone])
```

File: datetime/export_test.go (hunk 3 above)
Comment (oleg-jukovec) [reply datetime/export_test.go:1]: It's ok for export_test.go, see as examples:

https://go.dev/src/net/http/export_test.go
https://github.com/tarantool/go-tarantool/blob/master/export_test.go

File: datetime/example_test.go (hunk 2 above)
Comment (oleg-jukovec) [reply datetime/example_test.go]: GoLang by default sets UTC as timezone, unless you specify something else. So, yes. We need a value to specify a time without any timezone.

NoTimezone == "" (I've made a change) is a value for that purpose.

**Comment by oleg-jukovec** [comment]
I've made a separate pull request to fix golangci-lint error:
https://github.com/tarantool/go-tarantool/pull/204

We can ignore it for now.

**Review by DifferentialOrange — APPROVED** [review]

File: datetime/datetime_test.go [hunk 6]
```diff
@@ -73,6 +75,128 @@ func assertDatetimeIsEqual(t *testing.T, tuples []interface{}, tm time.Time) {
...
+		if offset != customOffset {
+			t.Fatalf("Expected offset %d instead of %d", customOffset, offset)
+		}
+
+		_, err = conn.Delete(spaceTuple1, 0, []interface{}{dt})
+		if err != nil {
+			t.Fatalf("Datetime delete failed: %s", err.Error())
+		}
+	} else {
+		t.Fatalf("datetime doesn't match")
```
Comment (vr009) [inline datetime/datetime_test.go:195]: Why do one the messages start with uppercase and another one with the lowercase?

File: datetime/datetime.go [hunk 7]
```diff
@@ -79,16 +77,44 @@ type Datetime struct {
...
 
 	if seconds < minSeconds || seconds > maxSeconds {
 		return nil, fmt.Errorf("Time %s is out of supported range.", t)
 	}
 
+	zone, offset := t.Zone()
+	if zone != NoTimezone {
+		if _, ok := timezoneToIndex[zone]; !ok {
+			return nil, fmt.Errorf("Unknown timezone %s with offset %d",
+				zone, offset)
```
Comment (vr009) [inline datetime/datetime.go:110]: Why do we use capitalized error strings? I have found some reasonable explanation why should not we do it. (See [this page](https://github.com/golang/go/wiki/CodeReviewComments#error-strings))

**Review by vr009 — APPROVED** [review]
LGTM. Please, check my comments.

File: datetime/datetime.go (hunk 7 above)
Comment (vr009) [reply datetime/datetime.go:110]: If you are agree with that point, we would need some code health patch for fixing it.

File: datetime/datetime_test.go (hunk 6 above)
Comment (oleg-jukovec) [reply datetime/datetime_test.go]: Thank you, fixed!

File: datetime/datetime.go (hunk 7 above)
Comment (oleg-jukovec) [reply datetime/datetime.go]: It's a good idea! I will fix all errors messages in a separate pull request when I will have a free time.

I've added a commit to fix messages format for the `datetime` package here.

Comment (vr009) [reply datetime/datetime.go]: Thank you!

**Pull request merged**